### PR TITLE
download Icon added in toolbar

### PR DESCRIPTION
**Pull request by iamutkarshtiwari**
@samdroid-apps This PR integrated this requested feature -> https://bugs.sugarlabs.org/ticket/4721
desktop-animation

### CODE REVIEW

**Comment by quozl** [comment]
Requires toolkit change https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/309

Which of the issues in the ticket are fixed?

Will you be updating the ticket?

Reviewed;
* please remove debug prints.

**Comment by iamutkarshtiwari** [comment]
@quozl Debug prints removed.

**Comment by iamutkarshtiwari** [comment]
@quozl I am afraid I won't be able to update the ticket since I haven't received the confirmation mail from bugs.sugarlabs.org. What would you suggest?

**Comment by quozl** [comment]
@walterbender on sugar-devel@ mailing list can do your confirmation if you send him the username you used on bugs.sugarlabs.org.

**Comment by samdroid-apps** [comment]
@iamutkarshtiwari Please try resending the activation email from trac.  I think we fixed it now.

Anyway, have you see how the download button works in epiphany (GNOME web).  I think it is really nice and that we should clone it.  It goes like:

* When you open epiphany, it is invisible
* When you start a download it becomes visible
[https://gyazo.com/020be0818910df5f7071d5d2b6c1881a](https://gyazo.com/020be0818910df5f7071d5d2b6c1881a)
* When click it, the palette shows a list of current downloads:
  * Clear removes all downloads from the list
  * If the download is in progress, it has a progress bar below it
  * If a download is finished it offers to open it in the journal (clicking the icon will open in app)
  * When you open something, it is removed from the list
* When a download is finished, the button stays
* When the user clears the downloads from the list, the button disappears again.

Maybe we could use those schematics for showing/hiding the button?

Anyway, the progress toolitem looks beautiful!

**Comment by quozl** [comment]
This pull request as it stands solves one of the issues in ticket 4721, and for that reason I'm in favour of it being merged.  The discussion on the mailing list was around 23rd January.

The additional features that @samdroid-apps refers to could be a different pull request.  I'd like to see them too, but no reason to delay this pull request now.

File: webtoolbar.py [hunk 1]
```diff
@@ -587,6 +608,7 @@ def _update_navigation_buttons(self):
         self._forward.props.sensitive = can_go_forward
 
         is_webkit_browser = isinstance(self._browser, Browser)
+        self._link_add.props.sensitive = is_webkit_browser
```
Comment (samdroid-apps) [inline webtoolbar.py]: Please don't revert unrelated commits

File: webtoolbar.py [hunk 2]
```diff
@@ -649,8 +671,7 @@ def _set_loading(self, loading):
 
     def _set_sensitive(self, value):
         for widget in self.toolbar:
-            if widget not in (self._stop_button,
-                              self._link_add):
+            if widget != self._stop_button:
```
Comment (samdroid-apps) [inline webtoolbar.py]: Please don't revert unrelated commits

File: webtoolbar.py [hunk 3]
```diff
@@ -436,6 +446,17 @@ def __switch_page_cb(self, tabbed_view, page, page_num):
     def __page_added_cb(self, notebook, child, pagenum):
         self.entry._search_popdown()
 
+    def __download_running_cb(self):
+        #print 'se llama'
```
Comment (samdroid-apps) [inline webtoolbar.py]: Please don't include debug prints

File: webtoolbar.py [hunk 4]
```diff
@@ -436,6 +446,17 @@ def __switch_page_cb(self, tabbed_view, page, page_num):
...
+        progress = downloadmanager.overall_downloads_progress()
+        #print 'progress', progress
+        self._download_icon.update(progress)
+        if progress > 0.0:
+            self._download_icon.set_tooltip("Downloading.. " + str(progress*100) + "%")
+        else:
+            self._download_icon.set_tooltip("No active downloads")
```
Comment (samdroid-apps) [inline webtoolbar.py]: Please run this string through gettext (`_("No...")`)

File: webtoolbar.py [hunk 5]
```diff
@@ -436,6 +446,17 @@ def __switch_page_cb(self, tabbed_view, page, page_num):
     def __page_added_cb(self, notebook, child, pagenum):
         self.entry._search_popdown()
 
+    def __download_running_cb(self):
+        #print 'se llama'
+        progress = downloadmanager.overall_downloads_progress()
+        #print 'progress', progress
+        self._download_icon.update(progress)
+        if progress > 0.0:
+            self._download_icon.set_tooltip("Downloading.. " + str(progress*100) + "%")
```
Comment (samdroid-apps) [inline webtoolbar.py]: Please run this string through gettext.  Also, please use a format string (better for translators).

Also, can you display the progress without the `.0`?  Just like `72%` not `72.0%`?

**Comment by samdroid-apps** [comment]
Yeah, good idea @quozl.  I have added my review to this patch.

**Comment by quozl** [comment]
@samdroid-apps, thanks for highlighting those unexpected changes.  It reminds me that I didn't look at them in detail, because the commit message didn't say what was being changed and why.

@iamutkarshtiwari, please also rework your commit messages; it gives us a great cross-check between what you think you are doing and what we see you changing.

**Comment by iamutkarshtiwari** [comment]
@samdroid-apps @quozl Updated and rebased the commit.

**Comment by quozl** [comment]
@iamutkarshtiwari, thanks.

To test on physical (not virtual) hardware, I had to do this;

```python
try:
    from sugar3.graphics.progresstoolbutton import ProgressToolButton
except:
    from progresstoolbutton import ProgressToolButton
```

So please add an instance of progresstoolbutton.py to your patch, otherwise the next version of Browse after this patch is merged won't be able to run on latest and earlier versions of Sugar.  (We don't *need* to wait for Sugar 0.110 before merging this patch, *if* you do this).

Tested on Fedora 18 with Sugar 0.108.1.

Your commit message refers to "overall_update" but the name does not exist.

As a guide, a commit message should;

- explain briefly what the problem was,
- explain briefly how the problem is solved,

Also, the first line of the commit message should be a summary, with subsequent lines in detail.  If this were my patch, the commit message would look something like;

```
Show progress of downloads (#4721)

User may be unaware of downloads because they happen in background.
Add a downloads progress tool button to show activity, and overall progress.
Part of a fix to https://bugs.sugarlabs.org/ticket/4721
```

But if you copy and paste that I'll grumble.  It's my work.  Use your own words.  :grin:

**Comment by iamutkarshtiwari** [comment]
@quozl Thanks for guiding me on all these mistakes. I'll keep them in mind :)

I have added an instance of progresstoolbutton.py to this patch and standardized the commit message as directed.

**Comment by quozl** [comment]
Reviewed.

File: webtoolbar.py [hunk 6]
```diff
@@ -426,6 +437,19 @@ def __init__(self, tabbed_view, act):
...
         self._configure_toolbar()
 
+    def __download_running_cb(self):
+        '''
+        Updates the downloadIcon tooltip message and progress value.
+        '''
+        progress = downloadmanager.overall_downloads_progress()
+        self._download_icon.update(progress)
+        if progress > 0.0:
+            self._download_icon.set_tooltip(_("Downloading.. " +
```
Comment (samdroid-apps) [inline webtoolbar.py]: Just 1 thing before I can merge:  please use a format string here (eg "Downloading... {}%".format()), so that it is eaiser for the translator to eg, put the % before the downloading.

**Comment by iamutkarshtiwari** [comment]
@samdroid-apps  All set!

File: webtoolbar.py (hunk 6 above)
Comment (iamutkarshtiwari) [inline webtoolbar.py]: Updated as suggested.

**Comment by quozl** [comment]
Please rebase and force push, so we have just one patch to merge?

**Comment by iamutkarshtiwari** [comment]
Rebased and forced pushed.

**Comment by iamutkarshtiwari** [comment]
@samdroid-apps I think we are good to go now?

**Comment by samdroid-apps** [comment]
I tested it agian.  Did you find a fix to the toolbutton size issue?  It is still taking up too much space :(  I was using SUGAR_SCALING=100 and it was around 3 grid cells wide (not the usual 1).

Otherwise, it is great on all the other fronts!

**Comment by iamutkarshtiwari** [comment]
@samdroid-apps I tried to find a way to fix that. I even tried re-patching my ProgressToolbutton module..but it seems that a bit-enhanced size of the icon is due to cairo-stroke over it. What shall I do now ? :(

**Comment by iamutkarshtiwari** [comment]
@samdroid-apps Please guide me on this so that we can get it done soon ;)

**Comment by iamutkarshtiwari** [comment]
@quozl Could you please guide me fix the icon scaling issue which Mr. Sam mentioned?

**Comment by iamutkarshtiwari** [comment]
@samdroid-apps @quozl Gotcha! I found the reason for the ```download icon size``` issue at sugar_scaling = 100 ;) It's because the 'emblem-download' icon's width is 56 pixels which is more than the normal width (55 px) of the other sugar-artwork icons. I guess i need to resize the 'emblem-download' icon to 55x55 and then push it to the sugar-artwork repository ?

**Comment by iamutkarshtiwari** [comment]
@samdroid-apps I see a little ambiguity. In your screenshot, the download icon is between 'address bar' and the navigation buttons.
sam

Whereas in my case the 'download icon' appears ```right``` to the navations buttons.
untitled

How could this be possible?

**Comment by samdroid-apps** [comment]
Hello,

The swapping of positions is weird, but I wouldn't be too worried about it.

Did you find a solution to the sizing issue?  If so, do I need an additional patch to those included in this pull request?

Thanks,
Sam

**Comment by iamutkarshtiwari** [comment]
@samdroidapps

**Comment by iamutkarshtiwari** [comment]
@samdroid-apps Could you please re-test it and let me know if you are still facing the same sizing issue?

There could be two possibilities-
1- The download icon ('emblem-download' icon) resolution is (```56```x55) unlike other regular icons' size (```55```x55) which might cause it to be wide.
2- The swapping of positions in your case might be causing that weird behaviour in width.

P.S.- I accidentally clicked the close PR button.

**Comment by iamutkarshtiwari** [comment]
@samdroid-apps I would suggest you re-test the latest commit I made because the tooltip in your screenshot appears ```behind``` the icon (might be because you were testing an older commit) unlike my screenshot (where the tooltip  appears ```around``` the icon).

**Comment by samdroid-apps** [comment]
On the latest version, it appears after the forward back buttons for me.  (I'm fine with that)

Did you make a patch for the issue with the icon size.

**Comment by iamutkarshtiwari** [comment]
Nopes.. Not yet. I am still not able to figure out why it's happening in your case? Could you please me a new screenshot of the latest patch so that I can an idea how it looks now.?

**Comment by davelab6** [comment]
@iamutkarshtiwari any progress on this? :)

**Comment by iamutkarshtiwari** [comment]
@davelab6 This PR works perfectly fine on my and @quozl's side, but there appears to be some problem at @samdroid-apps' end. Could you please test this PR and tell us if you are also facing the same problem as that of Sam's ?

**Comment by davelab6** [comment]
@iamutkarshtiwari I'm don't enough experience to test this yet, but I emailed sugar-devel to ask if anyone else can help :)

**Comment by samdroid-apps** [comment]
Great.  I'll merge it into the webkit2 branch.  It still is overflowing the size for me, but I'll debug it as it only seems to affect me :)

**Comment by iamutkarshtiwari** [comment]
@samdroid-apps Thank you :)

**Comment by samdroid-apps** [comment]
Ok, so I found the cause of the bug.

It happens when you change the tooltip.  Changing the tooltip calls Gtk.ToolButton.set_label.  That seems to think; "ok, I'm visible but I don't have a widget visible (for normal toolbuttons, it is a sugar3.graphics.icon.Icon widget).  I'll just make a Gtk.Button with the tooltip as the label and use that as my icon".  As the user, you don't see the icon, because the ProgressIcon has an overridden draw function.  However the button still affects the size allocate functions.

Fix; it seems to work fine if I '.set_icon_widget(ProgressIcon(...))'.  That also cuts the LoC down a lot, which is always a plus.

**Comment by iamutkarshtiwari** [comment]
@samdroid-apps Hi! What's the update on this PR?

**Comment by samdroid-apps** [comment]
I had a go at it:  https://github.com/samdroid-apps/browse-activity/commit/e86526ebb372b66ab03422e76e84cd539704e487

But it caused an issue with the palette tooltip (see commit description).  Do you see anything to help fix that?

I merged your commit into the webkit2 branch though.

**Comment by iamutkarshtiwari** [comment]
I tested this patch on an XO-4 and it still works fine.

**Comment by samdroid-apps** [comment]
Maybe have a look under Gtk+ 3.20?  (eg. fedora 24 VM)
